lib_static/open_project/authentication/strategies/warden/doorkeeper_oauth.rb

```diff
@@ -4,65 +4,60 @@ module OpenProject
   module Authentication
     module Strategies
       module Warden
-        ##
-        # Allows testing authentication via doorkeeper OAuth2 token
-        #
         class DoorkeeperOAuth < ::Warden::Strategies::Base
+          include FailWithHeader
+
           def valid?
-            ::Doorkeeper::OAuth::Token
-              .from_request(decorated_request, *Doorkeeper.configuration.access_token_methods)
-              .present?
+            access_token = ::Doorkeeper::OAuth::Token
+                             .from_request(decorated_request, *Doorkeeper.configuration.access_token_methods)
+            access_token.present? &&
+              begin
+                JWT.decode(access_token, nil, false)
+                false
+              rescue JWT::DecodeError
+                true
+              end
           end
 
           def authenticate!
-            token = ::Doorkeeper::OAuth::Token.authenticate(decorated_request,
-                                                            *Doorkeeper.configuration.access_token_methods)
-            return fail!("invalid token") unless token&.acceptable?(scope)
-
-            if token.resource_owner_id.nil?
-              authenticate_client_credentials(token)
+            access_token = ::Doorkeeper::OAuth::Token.authenticate(decorated_request,
+                                                                *Doorkeeper.configuration.access_token_methods)
+            if access_token.present?
+              if !access_token.expired? && !access_token.revoked?
+                if access_token.includes_scope?(scope)
+                  if access_token.resource_owner_id.nil?
+                    user_id = ::Doorkeeper::Application
+                                .where(id: access_token.application_id)
+                                .where.not(client_credentials_user_id: nil)
+                                .pick(:client_credentials_user_id)
+                    authenticate_user(user_id) if user_id
+                  else
+                    authenticate_user(access_token.resource_owner_id)
+                  end
+                else
+                  fail_with_header!(error: "insufficient_scope")
+                end
+              else
+                fail_with_header!(error: "invalid_token")
+              end
             else
-              authenticate_user(token.resource_owner_id)
+              fail_with_header!(error: "invalid_token")
```

Comment:
🟡 To reduce nesting, I normally prefer a coding style like:

```ruby
return fail_with_header!(error: "invalid_token") unless access_token.present?

return fail_with_header!(error: "invalid_token") if access_token.expired? || access_token.revoked?

return fail_with_header!(error: "insufficient_scope") if access_token.excludes_scope?(scope)

if access_token.resource_owner_id.nil?
  user_id = ::Doorkeeper::Application
              .where(id: access_token.application_id)
              .where.not(client_credentials_user_id: nil)
              .pick(:client_credentials_user_id)
  authenticate_user(user_id) if user_id
else
  authenticate_user(access_token.resource_owner_id)
end
```

WDYT? I find this much more readable.

Reply:
I think it looks nice. I will rewrite it. Thanks for suggestion!